File: parser/src/lev.ts
import { promisify } from 'util'
import { open, close, read, readFile } from 'fs'
import { readWadFd } from './wad'
import path from 'path'

const openAsync = promisify(open)
const readAsync = promisify(read)
const readFileAsync = promisify(readFile)
const closeAsync = promisify(close)

interface Lev {
    headerSize: number,
    version: number,
    obsoleteOffset: number,
    navigationOffset: number,
    uniqueIdCount: BigInt,
    width: number,
    height: number,
    mapHeaderSize: number,
    mapVersion: number,
    heightmapPalette: Buffer,
    ambientSoundVersion: number,
    soundThemesCount: number,
    soundPalette: Buffer,
    soundThemes: string[],
    checksum: number
    heightmap: LevHeightmapCell[],
    soundmap: LevSoundmapCell[]
}

interface LevHeightmapCell {
    size: number,
    version: number,
    height: number,
    groundTheme: Buffer,
    groundThemeStrength: Buffer,
    walkable: boolean,
    passover: boolean,
    soundTheme: number,
    shore: boolean
}

interface LevSoundmapCell {
    size: number,
    version: number,
    soundTheme: Buffer,
    soundThemeStrength: Buffer,
    soundIndex: number
}

/**
 * Parse Lev from inside a Wad file.
 */

export async function readLevFromWadFile (wadFile: string, levFile: string): Promise<Lev> {
    const fd = await openAsync(wadFile, 'r')

    let result = null

    try {
        result = await readLevFromWadFd(fd, levFile)
    } catch (error) {
        await closeAsync(fd)
        throw error
    }

    await closeAsync(fd)

    return result
}

export async function readLevFromWadFd (wadFd: number, levFile: string): Promise<Lev> {
    const wad = await readWadFd(wadFd)

    levFile = path.normalize(levFile.toLowerCase())

    const levEntry = wad.entries.find(file => path.normalize(file.name.toLowerCase()) === levFile)

    if (!levEntry) {
        throw new Error('Lev file ' + levEntry + ' does not exist')
    }

    const levData = Buffer.alloc(levEntry.length)
    await readAsync(wadFd, levData, 0, levEntry.length, levEntry.offset)

    return readLevBuffer(levData)
}

/**
 * Parse an extracted Lev file.
 */

export async function readLevFile (levFile: string): Promise<Lev> {
    return readLevBuffer(await readFileAsync(levFile))
}

/**
 * Parse Lev buffer.
 */

export function readLevBuffer (levData: Buffer): Lev {
    let offset = -4

    const headerSize = levData.readUInt32LE(offset += 4)
    const version = levData.readUInt16LE(offset += 2)

    // 3 unknwon byte
    offset += 3
    // 4 unknown bytes
    offset += 4

    // Could be useful?
    const obsoleteOffset = levData.readUInt32LE(offset += 4)

    // 4 unknown bytes. could be a uint32 like before
    offset+= 4

    const navigationOffset = levData.readUInt32LE(offset += 4)

    // file header ends and map header starts

    const mapHeaderSize = levData.readUInt8(offset += 1)
    const mapVersion = levData.readUInt32BE(offset += 4)

console.log('headerSize', headerSize)
console.log('version', version)

    // Electron doesn't support this yet
    // const uniqueIdCount = header.readBigUInt64LE(30)
    const uniqueIdCount = (BigInt(levData.readUInt32LE(offset += 4)) << 32n) | BigInt(levData.readUInt32LE(offset += 4))

    const width = levData.readUInt32LE(offset += 4)
    const height = levData.readUInt32LE(offset += 4)

    // This is a thing. Useless
    // const alwaysTrue = header.readUInt8LE(46)
    offset += 1

    // Theme data
    const heightmapPalette = levData.slice(offset, offset += 33792)
    const ambientSoundVersion = levData.readUInt32LE(offset += 4)
    const soundThemesCount = levData.readUInt32LE(offset += 4)
    const soundPalette = levData.slice(offset, offset += 33792)
    const checksum = levData.readUInt32LE(offset += 4)

    console.log('soundThemesCount', soundThemesCount)

    console.log('offset', offset)

    offset += 4

    // Sound themes
    let soundThemesIter = soundThemesCount - 1
    const soundThemes = []

    while (soundThemesIter--) {
        const length = levData.readUInt32LE(offset)
        const name = levData.slice(offset += 4, offset += length)

        console.log(length, name.slice(0, 100).toString())

        soundThemes.push(name.toString())
    }

    console.log('soundThemes', soundThemes)

    // console.log(soundThemes)

    // Heightmap
    let heightmapIter = (width + 1) * (height + 1)
    const heightmap: LevHeightmapCell[] = []

    while (heightmapIter--) {
        const size = levData.readUInt32LE(offset)
        const version = levData.readUInt8(offset + 4)
        const height = levData.readFloatLE(offset + 5)
        // const zero = levData.readUInt32LE(offset + 9)
        const groundTheme = levData.slice(offset + 10, offset + 13)
        const groundThemeStrength = levData.slice(offset + 13, offset + 15)
        const walkable = !!levData.readUInt8(offset + 15)
        const passover = !!levData.readUInt8(offset + 16)
        const soundTheme = levData.readUInt8(offset + 17)
        // const zero = levData.readUint8(offset + 18)
        const shore = !!levData.readUInt8(offset + 19)
        // const unknown = levData.readUint8(offset + 20)

        offset += 21

        heightmap.push({
            size,
            version,
            height,
            groundTheme,
            groundThemeStrength,
            walkable,
            passover,
            soundTheme,
            shore
        })
    }

    // console.log('offset', offset)
    // console.log('heightmap length', heightmap.length)
    // console.log('heightmap start', heightmap.slice(0, 5))
    // console.log('heightmap end', heightmap.slice(-5))

    // Sound Map
    // let soundMapIter = width * height
    let soundMapIter = 3328
    const soundmap: LevSoundmapCell[] = []

    while (soundMapIter--) {
        const size = levData.readUInt32LE(offset += 4)
        const version = levData.readUInt8(offset += 1)
        const soundTheme = levData.slice(offset, offset += 3)
        const soundThemeStrength = levData.slice(offset, offset += 2)
        const soundIndex = levData.readUInt8(offset += 1)

        soundmap.push({
            size,
            version,
            soundTheme,
            soundThemeStrength,
            soundIndex
        })
    }

    // console.log(soundmap.length, soundmap)

    console.log('offsets', offset, navigationOffset)

    return {
        headerSize,
        version,
        obsoleteOffset,
        navigationOffset,
        uniqueIdCount,
        width,
        height,
        mapHeaderSize,
        mapVersion,
        heightmapPalette,
        ambientSoundVersion,
        soundThemesCount,
        soundPalette,
        checksum,
        soundThemes,
        heightmap,
        soundmap,
    }
}

File: parser/src/lev.rs
use nom::IResult;
use nom::number::complete::{le_u8,le_u16,le_u32,le_u64,float};
use nom::bytes::complete::{tag,take};
use nom::sequence::tuple;
use nom::combinator::iterator;
use nom::multi::count;
use std::fs::{File,create_dir_all};
use std::io::{SeekFrom,Seek,Read,Write,Error,ErrorKind};
use std::iter::Iterator;
use std::collections::{HashMap,HashSet};
use std::path::Path;
use std::convert::TryInto;
use nom::branch::alt;

#[derive(Debug,PartialEq)]
pub struct Lev<'a> {
    header: LevHeader<'a>,
    heightmap_cells: Vec<LevHeightmapCell>,
    soundmap_cells: Vec<LevSoundmapCell>,
}

#[derive(Debug,PartialEq)]
pub struct LevHeader<'a> {
    pub version: u16,
    pub obsolete_offset: u32,
    pub navigation_offset: u32,
    pub unique_id_count: u64,
    pub width: u32,
    pub height: u32,
    pub map_version: u32,
    pub heightmap_palette: &'a [u8],
    pub ambient_sound_version: u32,
    pub sound_palette: &'a [u8],
    pub checksum: u32,
    pub sound_themes: Vec<&'a [u8]>,
}

pub fn parse_header(input: &[u8]) -> IResult<&[u8], LevHeader> {
    let (input, _header_size) = le_u32(input)?;
    let (input, version) = le_u16(input)?;
    let (input, _unknown_1) = take(3usize)(input)?;
    let (input, _unknown_2) = le_u32(input)?;
    let (input, obsolete_offset) = le_u32(input)?;
    let (input, _unknown_3) = le_u32(input)?;
    let (input, navigation_offset) = le_u32(input)?;
    let (input, _map_header_size) = le_u8(input)?;
    let (input, map_version) = le_u32(input)?;
    let (input, unique_id_count) = le_u64(input)?;
    let (input, width) = le_u32(input)?;
    let (input, height) = le_u32(input)?;
    let (input, _always_true) = le_u8(input)?;

    let (input, heightmap_palette) = take(33792usize)(input)?;
    let (input, ambient_sound_version) = le_u32(input)?;
    let (input, sound_themes_count) = le_u32(input)?;
    let (input, sound_palette) = take(33792usize)(input)?;
    let (input, checksum) = le_u32(input)?;

    let (input, sound_themes) = count(parse_header_sound_theme, (sound_themes_count - 1) as usize)(input)?;

    Ok(
        (
            input,
            LevHeader {
                version: version,
                obsolete_offset: obsolete_offset,
                navigation_offset: navigation_offset,
                unique_id_count: unique_id_count,
                width: width,
                height: height,
                map_version: map_version,
                heightmap_palette: heightmap_palette,
                ambient_sound_version: ambient_sound_version,
                sound_palette: sound_palette,
                checksum: checksum,
                sound_themes: sound_themes,
            }
        )
    )
}

pub fn parse_header_sound_theme(input: &[u8]) -> IResult<&[u8], &[u8]> {
    let (input, size) = le_u32(input)?;
    let (input, sound_theme) = take(size as usize)(input)?;
    Ok((input, sound_theme))
}

#[derive(Debug,PartialEq)]
pub struct LevHeightmapCell {
    size: u32,
    version: u8,
    height: f32,
    ground_theme: (u8, u8, u8),
    ground_theme_strength: (u8, u8),
    walkable: bool,
    passover: bool,
    sound_theme: u8,
    shore: bool,
}

pub fn parse_heightmap_cell(input: &[u8]) -> IResult<&[u8], LevHeightmapCell> {
    let (input, size) = le_u32(input)?;
    let (input, version) = le_u8(input)?;
    let (input, height) = float(input)?;
    let (input, _zero) = le_u32(input)?;
    let (input, ground_theme) = tuple((le_u8, le_u8, le_u8))(input)?;
    let (input, ground_theme_strength) = tuple((le_u8, le_u8))(input)?;
    let (input, walkable) = le_u8(input)?;
    let (input, passover) = le_u8(input)?;
    let (input, sound_theme) = le_u8(input)?;
    let (input, _zero) = le_u8(input)?;
    let (input, shore) = le_u8(input)?;
    let (input, _unknown) = le_u8(input)?;

    Ok(
        (
            input,
            LevHeightmapCell {
                size: size,
                version: version,
                height: height,
                ground_theme: ground_theme,
                ground_theme_strength: ground_theme_strength,
                walkable: walkable != 0,
                passover: passover != 0,
                sound_theme: sound_theme,
                shore: shore != 0,
            }
        )
    )
}

#[derive(Debug,PartialEq)]
pub struct LevSoundmapCell {
    size: u32,
    version: u8,
    sound_theme: (u8, u8, u8),
    sound_theme_strength: (u8, u8),
    sound_index: u8,
}

pub fn parse_soundmap_cell(input: &[u8]) -> IResult<&[u8], LevSoundmapCell> {
    let (input, size) = le_u32(input)?;
    let (input, version) = le_u8(input)?;
    let (input, sound_theme) = tuple((le_u8, le_u8, le_u8))(input)?;
    let (input, sound_theme_strength) = tuple((le_u8, le_u8))(input)?;
    let (input, sound_index) = le_u8(input)?;

    Ok(
        (
            input,
            LevSoundmapCell {
                size: size,
                version: version,
                sound_theme: sound_theme,
                sound_theme_strength: sound_theme_strength,
                sound_index: sound_index,
            }
        )
    )
}

#[derive(Debug,PartialEq)]
pub struct LevNavigationHeader<'a> {
    sections_start: u32,
    sections_count: u32,
    sections: Vec<(&'a [u8], u32)>,
}

fn parse_navigation_header(input: &[u8]) -> IResult<&[u8], LevNavigationHeader> {
    let (input, sections_start) = le_u32(input)?;
    let (input, sections_count) = le_u32(input)?;

    let (input, sections) = count(parse_navigation_header_section, sections_count as usize)(input)?;

    Ok(
        (
            input,
            LevNavigationHeader {
                sections_start: sections_start,
                sections_count: sections_count,
                sections: sections,
            }
        )
    )
}

fn parse_navigation_header_section(input: &[u8]) -> IResult<&[u8], (&[u8], u32)> {
    let (input, len) = le_u32(input)?;
    let (input, name) = take(len as usize)(input)?;
    let (input, start) = le_u32(input)?;

    Ok( (input, (name, start)) )
}

#[derive(Debug,PartialEq)]
pub struct LevNavigationSection {
    size: u32,
    version: u32,
    level_width: u32,
    level_height: u32,
    interactive_nodes: Vec<LevInteractiveNode>,
    subsets_count: u32,
    level_nodes: Vec<LevNavigationNode>,
}

fn parse_navigation_section(input: &[u8]) -> IResult<&[u8], LevNavigationSection> {
    let (input, size) = le_u32(input)?;
    let (input, version) = le_u32(input)?;
    let (input, level_width) = le_u32(input)?;
    let (input, level_height) = le_u32(input)?;
    let (input, _unknown_1) = le_u32(input)?;

    let (input, interactive_nodes_count) = le_u32(input)?;
    let (input, interactive_nodes) = count(parse_navigation_interactive_node, interactive_nodes_count as usize)(input)?;

    let (input, subsets_count) = le_u32(input)?;

    let (input, level_nodes_count) = le_u32(input)?;
    let (input, level_nodes) = count(parse_navigation_level_node, level_nodes_count as usize)(input)?;

    Ok(
        (
            input,
            LevNavigationSection {
                size: size,
                version: version,
                level_width: level_width,
                level_height: level_height,
                interactive_nodes: interactive_nodes,
                subsets_count: subsets_count,
                level_nodes: level_nodes,
            }
        )
    )
}

#[derive(Debug,PartialEq)]
struct LevInteractiveNode {
    x: u32,
    y: u32,
    subset: u32,
}

fn parse_navigation_interactive_node(input: &[u8]) -> IResult<&[u8], LevInteractiveNode> {
    let (input, x) = le_u32(input)?;
    let (input, y) = le_u32(input)?;
    let (input, subset) = le_u32(input)?;

    Ok(
        (
            input,
            LevInteractiveNode {
                x: x,
                y: y,
                subset: subset,
            }
        )
    )
}

#[derive(Debug,PartialEq)]
enum LevNavigationNode {
    Regular(LevNavigationRegularNode),
    Navigation(LevNavigationNavigationNode),
    Exit(LevNavigationExitNode),
    Blank(LevNavigationBlankNode),
}

fn parse_navigation_level_node(input: &[u8]) -> IResult<&[u8], LevNavigationNode> {
    alt((
        parse_navigation_regular_node,
        parse_navigation_navigation_node,
        parse_navigation_exit_node,
        parse_navigation_blank_node
    ))(input)
}

#[derive(Debug,PartialEq)]
struct LevNavigationRegularNode {
    root: u8,
    end: u8,
    layer: u8,
    subset: u8,
    x: f32,
    y: f32,
    node_id: u32,
    child_nodes: (u32, u32, u32, u32)
}

fn parse_navigation_regular_node(input: &[u8]) -> IResult<&[u8], LevNavigationNode> {
    let (input, _node_op) = tag(&[0, 0, 0, 0, 0, 1, 0, 0])(input)?;
    let (input, _unknown_1) = le_u8(input)?;
    let (input, root) = le_u8(input)?;
    let (input, _unknown_2) = le_u8(input)?;
    let (input, end) = le_u8(input)?;
    let (input, layer) = le_u8(input)?;
    let (input, subset) = le_u8(input)?;
    let (input, x) = float(input)?;
    let (input, y) = float(input)?;
    let (input, node_id) = le_u32(input)?;
    let (input, child_nodes) = tuple((le_u32, le_u32, le_u32, le_u32))(input)?;

    Ok(
        (
            input,
            LevNavigationNode::Regular(
                LevNavigationRegularNode {
                    root: root,
                    end: end,
                    layer: layer,
                    subset: subset,
                    x: x,
                    y: y,
                    node_id: node_id,
                    child_nodes: child_nodes,
                }
            )
        )
    )
}

#[derive(Debug,PartialEq)]
struct LevNavigationNavigationNode {
    root: u8,
    end: u8,
    layer: u8,
    subset: u8,
    x: f32,
    y: f32,
    node_id: u32,
    node_level: u32,
    nodes: Vec<u32>,
}

fn parse_navigation_navigation_node(input: &[u8]) -> IResult<&[u8], LevNavigationNode> {
    let (input, _node_op) = tag(&[0, 0, 0, 1, 0, 1, 0, 1])(input)?;
    let (input, _unknown_1) = le_u8(input)?;
    let (input, root) = le_u8(input)?;
    let (input, _unknown_2) = le_u8(input)?;
    let (input, end) = le_u8(input)?;
    let (input, layer) = le_u8(input)?;
    let (input, subset) = le_u8(input)?;
    let (input, x) = float(input)?;
    let (input, y) = float(input)?;
    let (input, node_id) = le_u32(input)?;
    let (input, node_level) = le_u32(input)?;
    let (input, _unknown_3) = le_u8(input)?;

    let (input, nodes_count) = le_u32(input)?;
    let (input, nodes) = count(le_u32, nodes_count as usize)(input)?;

    Ok(
        (
            input,
            LevNavigationNode::Navigation(
                LevNavigationNavigationNode {
                    root: root,
                    end: end,
                    layer: layer,
                    subset: subset,
                    x: x,
                    y: y,
                    node_id: node_id,
                    node_level: node_level,
                    nodes: nodes,
                }
            )
        )
    )
}

#[derive(Debug,PartialEq)]
struct LevNavigationExitNode {
    root: u8,
    end: u8,
    layer: u8,
    subset: u8,
    x: f32,
    y: f32,
    node_id: u32,
    node_level: u32,
    nodes: Vec<u32>,
    uids: Vec<u64>,
}

fn parse_navigation_exit_node(input: &[u8]) -> IResult<&[u8], LevNavigationNode> {
    let (input, _node_op) = tag(&[1, 0, 0, 1, 1, 0, 1, 1])(input)?;
    let (input, _unknown_1) = le_u8(input)?;
    let (input, root) = le_u8(input)?;
    let (input, _unknown_2) = le_u8(input)?;
    let (input, end) = le_u8(input)?;
    let (input, layer) = le_u8(input)?;
    let (input, subset) = le_u8(input)?;
    let (input, x) = float(input)?;
    let (input, y) = float(input)?;
    let (input, node_id) = le_u32(input)?;
    let (input, node_level) = le_u32(input)?;
    let (input, _unknown_3) = le_u8(input)?;

    let (input, nodes_count) = le_u32(input)?;
    let (input, nodes) = count(le_u32, nodes_count as usize)(input)?;

    let (input, uids_count) = le_u32(input)?;
    let (input, uids) = count(le_u64, uids_count as usize)(input)?;

    Ok(
        (
            input,
            LevNavigationNode::Exit(
                LevNavigationExitNode {
                    root: root,
                    end: end,
                    layer: layer,
                    subset: subset,
                    x: x,
                    y: y,
                    node_id: node_id,
                    node_level: node_level,
                    nodes: nodes,
                    uids: uids,
                }
            )
        )
    )
}

#[derive(Debug,PartialEq)]
struct LevNavigationBlankNode {
    root: u8
}

fn parse_navigation_blank_node(input: &[u8]) -> IResult<&[u8], LevNavigationNode> {
    let (input, _node_op) = tag(&[0, 1, 1])(input)?;
    let (input, _unknown_1) = le_u8(input)?;
    let (input, root) = le_u8(input)?;
    let (input, _unknown_2) = le_u8(input)?;

    Ok(
        (
            input,
            LevNavigationNode::Blank(
                LevNavigationBlankNode {
                    root: root,
                }
            )
        )
    )
}

// #![cfg(test)]
// mod tests {
//     use super::*;

//     fn test_
// }